### ISSUE TITLE
SVG images fail to paint

### ISSUE THREAD

**Issue by HalosGhost** [eXeC64/imv#149]
## System

Distro: Archlinux
package: -git built from master (also present in the officially-distributed version 3.0)

Other details: librsvg is installed and was installed at build-time (for the from-source version; neither version for Arch officially lists a dependency on librsvg)

## Problem

When loading any SVG images, SDL returns ``BAD_IMAGE``, which causes the navigator to drop image from the list and move to the next one.

When all files in the list are SVGs, this results in `imv` closing very quickly with the following message (and with a successful status code):

```console
No input files left. Exiting.
```

## Steps to reproduce

1. build `imv` from-source (or install from the repositories)
2. navigate to a directory that contains an SVG (e.g., `first.svg`)
3. run `imv` on the SVG file (e.g., `imv first.svg`)

## Example case

Though all SVG files I have on my system (generated or hand-coded) fail with `imv`, I thought it might be useful to give an example input:

```svg
<?xml version="1.0" encoding="UTF-8"?>
<svg fill="currentColor" class="issue-icon" version="1.1" viewBox="0 0 16 16" xmlns="http://www.w3.org/2000/svg">
  <circle r="7" cx="8" cy="8" stroke="currentColor" stroke-width="2" fill="none" />
  <rect width="2" height="2" x="7" y="11" rx="0.5" />
  <rect width="2" height="7" x="7" y="3" rx="0.5" />
</svg>
```

The example above is a hand-coded SVG of mine that I made in response to an SVG optimization article. It passes validation by the w3 validator and is parsed / displays sanely in inkscape and every browser I have access to which supports SVG.

I am happy to provide any other details which may be useful.

**Comment by eXeC64** [comment]
Hi, sorry I missed you on irc. I'm seeing some strange behaviour on my end, sometimes the SVG loads fine, sometimes imv fails to load it, and sometimes I get weird crashes.

At the very least it seems there's some kind of undefined behaviour being hit with how imv handles this svg file.

It's possible on your end you're not compiling with rsvg enabled, from the latest master, if you run `imv -h`, do you see librsvg listed as one of the built-in backends?

**Comment by eXeC64** [comment]
Okay, looks like it's down to this line:
```c
snprintf(path, sizeof path, "file://%s", src->name);
```

When I first added rsvg, it wouldn't load unless I included the `file://` prefix, but now it won't load unless I remove it. My hunch is something changed inside gnome, [this page](https://developer.gnome.org/rsvg/stable/rsvg-RsvgHandle.html) suggests URI syntax is only supported if the VFS layer in gnome is enabled.

I need to investigate how to reliably provide a path to librsvg. If all else fails, imv could always try both and see what sticks.

**Comment by eXeC64** [comment]
Side issue: imv has terrible error reporting for when an image fails to load. This needs to improve to make it obvious why a file is being rejected.

**Comment by HalosGhost** [comment]
> sorry I missed you on irc

Not a problem! I was really only checking to see if it was something silly I was doing and didn't want to add to issue noise if it didn't warrant.

> It's possible on your end you're not compiling with rsvg enabled, from the latest master, if you run `imv -h`, do you see librsvg listed as one of the built-in backends?

I already wondered if that was the case. and double-checked:

```console
imv v3.0.0-71-ge278e28
See manual for usage information.
This version of imv has been compiled with the following backends:

Name: libRSVG
Description: SVG library developed by GNOME
Website: https://wiki.gnome.org/Projects/LibRsvg
License: Lesser GNU Public License

Name: FreeImage
Description: Open source image library supporting a large number of formats
Website: http://freeimage.sourceforge.net/
License: FreeImage Public License v1.0

Legal:
imv's full source code is published under the terms of the MIT
license, and can be found at https://github.com/eXeC64/imv

imv uses the inih library to parse ini files.
See https://github.com/benhoyt/inih for details.
inih is used under the New (3-clause) BSD license.
```

It does, indeed, list librsvg.

Looks like you may have nailed down the issue. Let me know if I can do something to help test or confirm.

**Comment by eXeC64** [comment]
> Let me know if I can do something to help test or confirm.

I also use Arch as my daily driver, so I'm afraid your test environment is pretty much identical to mine. Thanks for the offer though.

I'm going to have a play with imv in a few different vms, see how librsvg behaves on a couple of other distributions, and maybe an old release of Ubuntu.

**Comment by HalosGhost** [comment]
No worries. :)

Thank you for looking into this!

**Comment by vglfr** [comment]
Loading svg with an absolute path (`imv ~/example.svg`) consistently succeeds on my Arch. With relative path it consistently fails.

**Comment by vglfr** [comment]
Did the trick for me:
```
char buff[PATH_MAX];
char path[PATH_MAX+8];
realpath(src->name, buff);
snprintf(path, sizeof path, "file://%s", buff);
```

**Comment by HalosGhost** [comment]
I just confirmed @vglfr's test regarding absolute and relative paths.

Awesome find! I am very excited to have this fixed!

**Comment by HalosGhost** [comment]
@vglfr, care to open a PR (I can if you would like)?

**Comment by vglfr** [comment]
@HalosGhost Just give me a few days, solution above is quick and untested.

**Comment by mb720** [comment]
Is this fix going to be included in imv 3.1.0? I'm asking since I just ran into this bug with imv 3.0.0 on Arch Linux.

**Comment by eXeC64** [comment]
Yes, it will. I'm hoping to release v3.1.0 shortly.
